cmd/snap-confine/seccomp-support.c

```diff
@@ -50,720 +51,51 @@
-	else if (strncmp(uts_machine, "ppc64le", 7) == 0)
-		return SCMP_ARCH_PPC64LE;
-#endif
-#if defined (SCMP_ARCH_PPC64)
-	else if (strncmp(uts_machine, "ppc64", 5) == 0)
-		return SCMP_ARCH_PPC64;
-#endif
-#if defined (SCMP_ARCH_PPC)
-	else if (strncmp(uts_machine, "ppc", 3) == 0)
-		return SCMP_ARCH_PPC;
-#endif
-#if defined (SCMP_ARCH_S390X)
-	else if (strncmp(uts_machine, "s390x", 5) == 0)
-		return SCMP_ARCH_S390X;
-#endif
-	return 0;
-}
-
-static uint32_t get_hostarch(void)
-{
-	struct utsname uts;
-	if (uname(&uts) < 0)
-		die("uname() failed");
-	uint32_t arch = uts_machine_to_seccomp_arch(uts.machine);
-	if (arch > 0)
-		return arch;
-	// Just return the seccomp userspace native arch if we can't detect the
-	// kernel host arch.
-	return seccomp_arch_native();
-}
-
-static void sc_add_seccomp_archs(scmp_filter_ctx * ctx)
-{
-	uint32_t native_arch = seccomp_arch_native();	// seccomp userspace
-	uint32_t host_arch = get_hostarch();	// kernel
-	uint32_t compat_arch = 0;
-
-	debug("host arch (kernel) is '%d'", host_arch);
-	debug("native arch (userspace) is '%d'", native_arch);
-
-	// For architectures that support a compat architecture, when the
-	// kernel and userspace match, add the compat arch, otherwise add
-	// the kernel arch to support the kernel's arch (eg, 64bit kernels with
-	// 32bit userspace).
-	if (host_arch == native_arch) {
-		switch (host_arch) {
-#if defined (SCMP_ARCH_X86_64)
-		case SCMP_ARCH_X86_64:
-			compat_arch = SCMP_ARCH_X86;
-			break;
-#endif
-#if defined(SCMP_ARCH_AARCH64)
-		case SCMP_ARCH_AARCH64:
-			compat_arch = SCMP_ARCH_ARM;
-			break;
-#endif
-#if defined (SCMP_ARCH_PPC64)
-		case SCMP_ARCH_PPC64:
-			compat_arch = SCMP_ARCH_PPC;
-			break;
-#endif
-		default:
-			break;
-		}
-	} else
-		compat_arch = host_arch;
-
-	if (compat_arch > 0 && seccomp_arch_exist(ctx, compat_arch) == -EEXIST) {
-		debug("adding compat arch '%d'", compat_arch);
-		if (seccomp_arch_add(ctx, compat_arch) < 0)
-			die("seccomp_arch_add(..., compat_arch) failed");
-	}
-}
-
-scmp_filter_ctx sc_prepare_seccomp_context(const char *filter_profile)
-{
-	int rc = 0;
-	scmp_filter_ctx ctx = NULL;
-	FILE *f = NULL;
-	size_t lineno = 0;
 	uid_t real_uid, effective_uid, saved_uid;
-	struct preprocess pre;
-	struct seccomp_args sargs;
+	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
+		die("could not find user IDs");
+	// If not root but can raise, then raise privileges to load seccomp
+	// policy since we don't have nnp
+	debug("raising privileges to load seccomp profile");
+	if (effective_uid != 0 && saved_uid == 0) {
+		if (seteuid(0) != 0)
+			die("seteuid failed");
+		if (geteuid() != 0)
+			die("raising privs before seccomp_load did not work");
+	}
 
-	debug("preparing seccomp profile associated with security tag %s",
-	      filter_profile);
+	char profile_path[512];	// arbitrary path name limit
+	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s.bpf",
+			 filter_profile_dir, filter_profile);
```

Comment:
I will have a look at this, we may not need it anymore with real spread tests and with the bpf VM unit tests. But if its easier with it I will bring it back.